pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java

```diff
@@ -107,15 +101,21 @@ public Iterable<? extends ApexNode<?>> children() {
         return (Iterable<? extends ApexNode<?>>) super.children();
     }
 
-    abstract void calculateLineNumbers(SourceCodePositioner positioner);
+    void calculateLineNumbers(SourceCodePositioner positioner) {
```

Comment:
All of these classes are `abstract`. Do you mean leaving the methods `abstract` as well and instead providing these default implementations in `Empty`?

Reply:
Thanks, yes, that was what I intended to suggest. The reason being that if the derived class wants to inherit a set a common behaviors that is clear from the parent's name: `AbstractApexNode.Single`, `AbstractApexNode.Many`, `AbstractApexNode.Empty`. And then inheriting from just `AbstractApexNode` includes no assumptions about Summit<->PMD mapping, which will all be provided by the derived class.

Reply:
I have been thinking about this, and although I initially agreed, I'm not sure if it's the right way to go in all cases. For a node like [`ASTEmptyReferenceExpression`](https://github.com/eklimo/pmd/pull/8/commits/9c23644b22a165cd9e447d1f35ef76a8bf272542#diff-278ba4e72958bc9750552cf0bde82002d8d0338d836456651148a913be8e2b39), which neither wraps a node directly (type parameter) nor indirectly (`Node` fields), the name `Empty` seems appropriate. However, a node like [`ASTField`](https://github.com/eklimo/pmd/pull/11/commits/57b76534b1d44b714bc52f9240f69d01262e69cc#diff-981dc32a5c105a05091f9b379c3c93914f6202b30b758b6c06e0182773e1f050) does indirectly wrap nodes through its fields, so the name `Empty` seems less appropriate.

We could:
1. Add `Empty` for nodes like `ASTEmptyReferenceExpression`, but allow nodes like `ASTField` to extend `AbstractApexNode` directly.
2. Find a name instead of `Empty` that is appropriate for both cases. (I am not sure what that would be.)

Reply:
Right, the suggestion was that "special cases" (like `ASTField`) should inherit from `AbstractApexNode` directly and implement (e.g.) `ASTField.calculateLineNumbers` as appropriate, but that the base class methods e.g. `AbstractApexNode.calculateLineNumbers` should be abstract and unimplemented. These default implementations (like you have here, that make sense for "empty" methods) should be in a separate class `Empty` that would only be extended by nodes where it applies (like `ASTEmptyReferenceExpression`).

Reply:
Please see [`bb813f9`](https://github.com/eklimo/pmd/pull/8/commits/bb813f9701fdbb9a85f00a7db59409cc937e5b95)

Reply:
LGTM